modules/templates/eval/eval.go

```diff
@@ -0,0 +1,332 @@
+	sb.WriteString(err.msg)
+	sb.WriteString(" [ ")
+	for _, token := range err.tokens {
+		_, _ = fmt.Fprintf(&sb, `"%v" `, token)
+	}
+	sb.WriteString("]")
+	if err.err != nil {
+		sb.WriteString(": ")
+		sb.WriteString(err.err.Error())
+	}
+	return sb.String()
+}
+
+func (err ExprError) Unwrap() error {
+	return err.err
+}
+
+func (e *Eval) applyOp() {
+	op := e.stackOp.pop()
+	if op == "not" {
+		num := e.stackNum.pop()
+		i, _ := util.ToInt64(num.Value)
+		e.stackNum.push(Num{truth(i == 0)})
+	} else {
+		num2 := e.stackNum.pop()
+		num1 := e.stackNum.pop()
+		e.stackNum.push(applyOp2(op, num1, num2))
+	}
+}
+
+// Exec evaluates the given expression tokens and returns the result.
+// If no error occurs, the result is either an int64 or a float64.
+// If all numbers are integer, the result is an int64, otherwise if there is any float number, the result is a float64.
+// Golang's template syntax supports comparable int types: {{if lt $i32 $i64}} is right.
+func (e *Eval) Exec(tokens ...any) (ret Num, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			rErr, ok := r.(error)
+			if !ok {
+				rErr = fmt.Errorf("%v", r)
+			}
+			err = ExprError{"invalid expression", tokens, rErr}
+		}
+	}()
+	for _, token := range tokens {
+		n, err := toNum(token)
+		if err == nil {
+			e.stackNum.push(n)
+			continue
+		}
+
+		op, err := toOp(token)
+		if err != nil {
+			return Num{}, ExprError{"invalid expression", tokens, err}
+		}
+
+		switch {
+		case op.hasOpenBracket():
+			e.stackOp.push(op)
+		case op.isCloseBracket(), op.isComma():
+			var sop operator
+			for {
+				sop = e.stackOp.peek()
+				if sop.hasOpenBracket() || sop.isComma() {
+					break
+				}
+				e.applyOp()
+			}
+			if op.isCloseBracket() {
+				nums := []Num{e.stackNum.pop()}
+				for !e.stackOp.peek().hasOpenBracket() {
+					sop = e.stackOp.pop()
+					if !sop.isComma() {
+						return Num{}, ExprError{"bracket doesn't match", tokens, nil}
+					}
+					nums = append(nums, e.stackNum.pop())
+				}
+				for i, j := 0, len(nums)-1; i < j; i, j = i+1, j-1 {
+					nums[i], nums[j] = nums[j], nums[i] // reverse nums slice, to get the right order for arguments
+				}
+				sop = e.stackOp.pop()
+				fn := string(sop[:len(sop)-1])
+				if fn == "" {
+					if len(nums) != 1 {
+						return Num{}, ExprError{"too many values in one bracket", tokens, nil}
+					}
+					e.stackNum.push(nums[0])
+				} else if f, ok := e.funcMap[fn]; ok {
+					e.stackNum.push(f(nums))
+				} else {
+					return Num{}, ExprError{"unknown function: " + fn, tokens, nil}
+				}
+			} else {
+				e.stackOp.push(op)
+			}
+		default:
+			for len(e.stackOp.elems) > 0 && len(e.stackNum.elems) > 0 {
+				sop := e.stackOp.peek()
+				if sop.isComma() || precedence(sop, op) < 0 {
+					break
+				}
+				e.applyOp()
+			}
+			e.stackOp.push(op)
+		}
+	}
+	for len(e.stackOp.elems) > 0 {
+		e.applyOp()
+	}
+	if len(e.stackNum.elems) != 1 {
+		return Num{}, ExprError{"too many final values", tokens, nil}
```

Comment:
```suggestion
		return Num{}, ExprError{"too many values at the end", tokens, nil}
```
or something similar?
the `final` absolutely confuses me, it reads to me like `too many values that cannot be modified anymore`…

Reply:
How about `expect only one value for final result, but there are %d`